### PR TITLE
TASK: Add option to flush an optional workspace as well as the nodes workspace

### LINKED ISSUE #2224: The target workspace for published nodes is not flushed
If you publish a node the node's workspace content cache itself is flushed correctly. Publishing a node should also flush the content cache. This is done via
```php 
 $dispatcher->connect(PublishingService::class, 'nodePublished', ContentCacheFlusher::class, 'registerNodeChange');
```

unfortunately the methode signature for the signal `$this->emitNodePublished($node, $targetWorkspace)` passes the target workspace as an addtional argument wich is not respected on the corresponding slot `  public function registerNodeChange(NodeInterface $node)`.

The `registerNodeChange()` should respect this workspace as well!

### PR DESCRIPTION
**Pull request by johannessteu**
Resolves #2224

### CODE REVIEW

File: Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php [hunk 1]
```diff
@@ -92,17 +92,22 @@ class ContentCacheFlusher
...
         $this->tagsToFlush[ContentCache::TAG_EVERYTHING] = 'which were tagged with "Everything".';
 
         if (empty($this->workspacesToFlush[$node->getWorkspace()->getName()])) {
             $this->resolveWorkspaceChain($node->getWorkspace());
         }
 
+        if ($targetWorkspace !== null && empty($this->workspacesToFlush[$targetWorkspace->getName()])) {
+            $this->resolveWorkspaceChain($targetWorkspace);
+        }
+
```
Comment (kitsunet) [inline Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]: After we resolved the chain we also need to use it as in line 116 or am I mistaken? For now it looks like we resolve the additional chain but then do nothing with it.

Comment (johannessteu) [reply Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]: You're totally right!

**Comment by johannessteu** [comment]
Is there a better way to handle this signal information thingy? Those are per default true. We now introduced an optional argument. If it's not set within the signal it will be filled with signalInformations by default, wich is a string >.>
Doesn't really feel good to disable the flag on all signals

**Comment by kitsunet** [comment]
The signal info thing neesd a refactoring at some point...

**Comment by Nikdro** [comment]
@johannessteu Just tried that fix. Works for me. Thanks for fixing!

**Comment by kdambekalns** [comment]
See also https://github.com/neos/flow-development-collection/issues/1003

File: Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php [hunk 2]
```diff
@@ -92,23 +92,41 @@ class ContentCacheFlusher
...
+        if ($targetWorkspace !== null) {
+            $this->registerAllTagsToFlushForNodeInWorkspace($node, $targetWorkspace);
+        }
+    }
+
+    /**
+     * @param NodeInterface $node
+     * @param Workspace $workspace
+     */
+    protected function registerAllTagsToFlushForNodeInWorkspace(NodeInterface $node, Workspace $workspace)
```
Comment (daniellienert) [inline Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]: Nitpick: Would be nice to have complete return types for newly introduced methods - so void in this case

File: Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php [hunk 3]
```diff
@@ -92,23 +92,41 @@ class ContentCacheFlusher
...
+    {
         $nodeIdentifier = $node->getIdentifier();
-        foreach ($this->workspacesToFlush[$node->getWorkspace()->getName()] as $workspaceName => $workspaceHash) {
+
+        foreach ($this->workspacesToFlush[$workspace->getName()] as $workspaceName => $workspaceHash) {
```
Comment (daniellienert) [inline Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]: Should we check if key exists / is array before itterating over it?

File: Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php [hunk 4]
```diff
@@ -92,23 +92,41 @@ class ContentCacheFlusher
      * model or the Neos Publishing Service.
      *
      * @param NodeInterface $node The node which has changed in some way
+     * @param Workspace $targetWorkspace An optional workspace to flush
      * @return void
      * @throws \Neos\ContentRepository\Exception\NodeTypeNotFoundException
      */
-    public function registerNodeChange(NodeInterface $node)
+    public function registerNodeChange(NodeInterface $node, Workspace $targetWorkspace = null)
```
Comment (kdambekalns) [inline Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]:
```suggestion
    public function registerNodeChange(NodeInterface $node, Workspace $targetWorkspace = null): void
```

File: Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php (hunk 2 above)
Comment (kdambekalns) [reply Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]:
```suggestion
    protected function registerAllTagsToFlushForNodeInWorkspace(NodeInterface $node, Workspace $workspace): void
```

File: Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php (hunk 3 above)
Comment (kdambekalns) [reply Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php]: Why does `$this->workspacesToFlush` contain lists of workspaces for a workspace? I'm puzzled… anyways:

```suggestion
        if (!array_key_exists($workspace->getName(), $this->workspacesToFlush) || is_array($this->workspacesToFlush[$workspace->getName()]) === false) {
            return;
        }

        foreach ($this->workspacesToFlush[$workspace->getName()] as $workspaceName => $workspaceHash) {
```

**Pull request merged**